src/deploy/functions/release/planner.ts

```diff
@@ -10,39 +10,63 @@ export interface EndpointUpdate {
   deleteAndRecreate?: backend.Endpoint;
 }
 
-export interface RegionalChanges {
+export interface Changeset {
   endpointsToCreate: backend.Endpoint[];
   endpointsToUpdate: EndpointUpdate[];
   endpointsToDelete: backend.Endpoint[];
 }
 
-export type DeploymentPlan = Record<string, RegionalChanges>;
+export type DeploymentPlan = Record<string, Changeset>;
 
 export interface Options {
   filters?: string[][];
   // If set to false, will delete only functions that are managed by firebase
   deleteAll?: boolean;
 }
 
-/** Calculate the changes needed for a given region. */
+/** Calculate the changesets needed for a given region by grouping endpoints with keyFn. */
 export function calculateRegionalChanges(
   want: Record<string, backend.Endpoint>,
   have: Record<string, backend.Endpoint>,
+  keyFn: (e: backend.Endpoint) => string,
   options: Options
-): RegionalChanges {
-  const endpointsToCreate = Object.keys(want)
-    .filter((id) => !have[id])
-    .map((id) => want[id]);
-
-  const endpointsToDelete = Object.keys(have)
-    .filter((id) => !want[id])
-    .filter((id) => options.deleteAll || isFirebaseManaged(have[id].labels || {}))
-    .map((id) => have[id]);
-
-  const endpointsToUpdate = Object.keys(want)
-    .filter((id) => have[id])
-    .map((id) => calculateUpdate(want[id], have[id]));
-  return { endpointsToCreate, endpointsToUpdate, endpointsToDelete };
+): Record<string, Changeset> {
+  const toCreate = utils.groupBy(
+    Object.keys(want)
+      .filter((id) => !have[id])
+      .map((id) => want[id]),
+    keyFn
+  );
+
+  const toDelete = utils.groupBy(
+    Object.keys(have)
+      .filter((id) => !want[id])
+      .filter((id) => options.deleteAll || isFirebaseManaged(have[id].labels || {}))
```

Comment:
This doesn't seem like new behavior, but it looks like options.deleteAll deletes every function on the project, not just Firebase managed functions. Is that intentional?

Reply:
I think this option is only used in `functions:delete` where you can delete arbitrary function as long as you specify it's name correctly. I think that might be useful, though we might want to revisit that API.